### PR TITLE
dmctl: simplify default query-status output

### PR DESCRIPTION
**Pull request by lichunzhu**
### What problem does this PR solve? 
[TOOL-1632](https://internal.pingcap.net/jira/browse/TOOL-1632)
Current `query-status` command without any args in dmctl will [print a lot of information](https://pingcap.com/docs-cn/v3.0/reference/tools/data-migration/query-status/) but users are concerned about a little of them.

### What is changed and how it works?
This PR generates and returns a simple task-list when users **don't specify any arg**. The proto files are not changed. This PR only generates task-list info in **dmctl** and modifies the output after receiving full dm-workers' info from dm-master.

The output will be like this format:
```json
{
    "result": true,
    "msg: "",
    "tasks": [
        {
            "taskName": "task-1",
            "taskStatus": "Running",
            "workers": ["127.0.0.1:8262"]
        },
        {
            "taskName": "task-2",
            "taskStatus": "Error - Some error occurred in subtask",
            "workers": ["127.0.0.1:8262", "127.0.0.1:8263"]
        }
    ]
}
```
The status of a task is based on its subtasks. The definition is listed as follows:

|                     Subtasks' status                       |                Task's status                 |
| :--------------------------------------------------------: | :------------------------------------------: |
|           Any Paused and len(result.errors) > 0            |    Error - Some error occurred in subtask    |
| Any Running and unit is "Sync" and relay is Paused/Stopped | Error - Relay status is Error/Paused/Stopped |
|              Any Paused but without error                  |                    Paused                    |
|                        All New                             |                     New                      |
|                      All Finished                          |                   Finished                   |
|                      All Stopped                           |                   Stopped                    |
|                         Others                             |                   Running                    |
### Check List 

Tests 

 - Unit test

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: dm/ctl/master/query_status.go [hunk 1]
```diff
@@ -60,5 +74,49 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+		common.PrettyPrintResponse(resp)
+	}
+}
+
+// wrapTaskResult pick task info and generate tasks' status and relative workers
+func wrapTaskResult(resp *pb.QueryStatusListResponse) *taskResult {
+	taskStatusMap := make(map[string]string, len(resp.Workers))
+	taskCorrespondingWorkers := make(map[string][]string, len(resp.Workers))
+	for _, worker := range resp.Workers {
+		for _, subTask := range worker.SubTaskStatus {
```
Comment (csuzhangxc) [inline dm/ctl/master/query_status.go]: do we need to consider the status of _relay_ unit when _sync_ is the current process unit?

File: tests/start_task/run.sh [hunk 2]
```diff
@@ -37,8 +37,8 @@ function run() {
         check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT
 
         echo "check un-accessible DM-worker exists"
-        run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
-            "query-status" \
+		run_dm_ctl $WORK_DIR "127.0.0.1:$MASTER_PORT" \
```
Comment (csuzhangxc) [inline tests/start_task/run.sh]: mistake to convert `space` to `tab`?

File: dm/ctl/master/query_status.go [hunk 3]
```diff
@@ -60,5 +74,49 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+			subTaskName := subTask.Name
+			subTaskStage := subTask.Stage
+
+			taskCorrespondingWorkers[subTaskName] = append(taskCorrespondingWorkers[subTaskName], worker.Worker)
+			taskStage := taskStatusMap[subTaskName]
+			switch {
+			case taskStage == stageError:
+			case subTaskStage == pb.Stage_Paused && subTask.Result != nil && len(subTask.Result.Errors) > 0:
+				taskStatusMap[subTaskName] = stageError
+			case taskStage == pb.Stage_Paused.String():
```
Comment (csuzhangxc) [inline dm/ctl/master/query_status.go]: 1. do nothing for this `case`? if so, can we remove it?

2. should we add some comments to explain these stages here?

Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: 1. If this case is removed, taskStage might be set to "Running" although taskStage is "Paused".

2. I will add some comments.

**Comment by csuzhangxc** [comment]
@csuzhangxc @WangXiangUSTC PTAL

File: tests/start_task/run.sh (hunk 2 above)
Comment (lichunzhu) [reply tests/start_task/run.sh]: Revised

File: dm/ctl/master/query_status.go (hunk 1 above)
Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: The definition of task's status has been updated in PR description. The code has been revised correspondingly.

File: dm/ctl/master/query_status.go [hunk 4]
```diff
@@ -24,6 +24,20 @@ import (
...
+
+type taskResult struct {
+	Result bool        `json:"result"`
+	Msg    string      `json:"msg"`
+	Tasks  []*taskInfo `json:"tasks"`
+}
+
+type taskInfo struct {
+	TaskName   string   `json:"taskName,omitempty"`
+	TaskStatus string   `json:"taskStatus,omitempty"`
```
Comment (WangXiangUSTC) [inline dm/ctl/master/query_status.go]: how about use same format with others, changed to `task-status`? and also the task-name.

File: dm/ctl/master/query_status.go [hunk 5]
```diff
@@ -60,5 +74,75 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
-	common.PrettyPrintResponse(resp)
+	if resp.Result && taskName == "" && len(workers) == 0 {
+		result := wrapTaskResult(resp)
+		common.PrettyPrintInterface(result)
+	} else {
+		common.PrettyPrintResponse(resp)
+	}
+}
+
+// errorOccurred check ProcessResult and return true if some error occurred
```
Comment (WangXiangUSTC) [inline dm/ctl/master/query_status.go]:
```suggestion
// errorOccurred checks ProcessResult and return true if some error occurred
```

File: dm/ctl/master/query_status.go [hunk 6]
```diff
@@ -60,5 +74,75 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+
+// getRelayStage returns current relay stage (including stageError)
+func getRelayStage(relayStatus *pb.RelayStatus) string {
+	if errorOccurred(relayStatus.Result) {
+		return stageError
+	}
+	return relayStatus.Stage.String()
+}
+
+// wrapTaskResult pick task info and generate tasks' status and relative workers
```
Comment (WangXiangUSTC) [inline dm/ctl/master/query_status.go]:
```suggestion
// wrapTaskResult picks task info and generate tasks' status and relative workers
```

File: dm/ctl/master/query_status.go [hunk 7]
```diff
@@ -60,5 +74,75 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+			// | :--------------------------------------------------------: | :------------------------------------------: |
+			// |           Any Paused and len(result.errors) > 0            |    Error - Some error occurred in subtask    |
+			// | Any Running and unit is "Sync" and relay is Paused/Stopped | Error - Relay status is Error/Paused/Stopped |
+			// |              Any Paused but without error                  |                    Paused                    |
+			// |                        All New                             |                     New                      |
+			// |                      All Finished                          |                   Finished                   |
+			// |                      All Stopped                           |                   Stopped                    |
+			// |                         Others                             |                   Running                    |
+			switch {
+			case len(taskStage) >= 5 && taskStage[0:5] == stageError:
```
Comment (WangXiangUSTC) [inline dm/ctl/master/query_status.go]: what is the `5` means? need add some comment

Comment (csuzhangxc) [reply dm/ctl/master/query_status.go]: maybe you can use `strings.HasPrefix`?

File: dm/ctl/master/query_status.go (hunk 4 above)
Comment (csuzhangxc) [reply dm/ctl/master/query_status.go]: tag generated by protobuf is `CamelCase` :sweat_smile:

Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: maybe we should keep same with tag in other commands' output?

File: dm/ctl/master/query_status.go (hunk 7 above)
Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: Revised in https://github.com/pingcap/dm/pull/334/commits/e635e5163c72337b06ce1ec20b66f0b47be3487e

**Comment by lichunzhu** [comment]
The code has been revised. PTAL again. @csuzhangxc @WangXiangUSTC

File: dm/ctl/master/query_status_test.go [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+	resp.Workers[0].SubTaskStatus[0].Unit = 0
+	resp.Workers[0].RelayStatus = nil
+	expectedResult[0].TaskStatus = pb.Stage_Paused.String()
+	generateAndCheckTaskResult(c, resp, expectedResult)
+	// All subtasks are Finished
+	subTestSameSubTaskStatus(c, resp, expectedResult, pb.Stage_Finished)
+	// All subtasks are Stopped
+	subTestSameSubTaskStatus(c, resp, expectedResult, pb.Stage_Stopped)
+	// All subtasks are New
+	subTestSameSubTaskStatus(c, resp, expectedResult, pb.Stage_New)
```
[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/ctl/master/query_status_test.go]: Revised in https://github.com/pingcap/dm/pull/334/commits/00fe0862e45afa8ace0508b610199eca961bc911.

File: dm/ctl/master/query_status.go [hunk 9]
```diff
@@ -60,5 +75,75 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+			taskStage := taskStatusMap[subTaskName]
+			// the status of a task is decided by its subtasks, the rule is listed as follows:
+			// |                     Subtasks' status                       |                Task's status                 |
+			// | :--------------------------------------------------------: | :------------------------------------------: |
+			// |           Any Paused and len(result.errors) > 0            |    Error - Some error occurred in subtask    |
+			// | Any Running and unit is "Sync" and relay is Paused/Stopped | Error - Relay status is Error/Paused/Stopped |
+			// |              Any Paused but without error                  |                    Paused                    |
+			// |                        All New                             |                     New                      |
+			// |                      All Finished                          |                   Finished                   |
+			// |                      All Stopped                           |                   Stopped                    |
```
Comment (amyangfei) [inline dm/ctl/master/query_status.go]: seems we can't get `New/Finished/Stopped` status from current logic. Shall we assign the value to taskStatusMap for each first loop index?

File: dm/ctl/master/query_status.go [hunk 10]
```diff
@@ -60,5 +75,75 @@ func queryStatusFunc(cmd *cobra.Command, _ []string) {
...
+	if errorOccurred(relayStatus.Result) {
+		return stageError
+	}
+	return relayStatus.Stage.String()
+}
+
+// wrapTaskResult picks task info and generate tasks' status and relative workers
+func wrapTaskResult(resp *pb.QueryStatusListResponse) *taskResult {
+	taskStatusMap := make(map[string]string, len(resp.Workers))
+	taskCorrespondingWorkers := make(map[string][]string, len(resp.Workers))
```
Comment (amyangfei) [inline dm/ctl/master/query_status.go]: A little weird here, both `taskStatusMap` and `taskCorrespondingWorkers` are mapping from taskName to something, the capacity of map doesn't equal to `len(resp.Workers)`, use zero is ok

File: dm/ctl/master/query_status.go (hunk 9 above)
Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: `case taskStage == "", subTaskStage == pb.Stage_Paused` will be satisfied for each first loop index and the value of `subTaskStage` will be assigned to `taskStatusMap`.

File: dm/ctl/master/query_status.go (hunk 10 above)
Comment (lichunzhu) [reply dm/ctl/master/query_status.go]: Revised in https://github.com/pingcap/dm/pull/334/commits/6856319104013b88d10b25a3b53a5770ca781b7a.

**Pull request merged**